### ISSUE TITLE
Sign in with Drexel SSO

### ISSUE THREAD

**Issue by Zohair-coder** [Zohair-coder/drexel-scraper#89]
Drexel TMS used to be open for all students - anyone could access it. However, Drexel will soon implement SSO. We need to adapt the scraper to work with Drexel SSO. I got this information from the following email from Drexel IT:

```
Hi Zohair,

Thank you for your response and information.

Heads up - The Registrar’s Office made the decision to bring WebTMS behind SSO to be in line with compliance of campus security and safety to ensure only active schedule constituents can access the information online.

We have implemented their request in test. I don’t know when they will request it to move in production. They will send an email communication when it goes live in production.

If you have a need to test your code for SSO, you can check our test environment. You need to be within Drexel VPN to use test environment. Also keep in mind that test has limited assigned connections, it’s okay for debugging code, but not good for massive stress test.

https://termmasterschedule.irttest.drexel.edu/webtms_du/

Thanks,
Hannah

Drexel Internal Data
```

**Comment by Zohair-coder** [comment]
If doing this with simple HTTP requests proves to be too difficult, I am open to tackling this issue using tools like Selenium (https://selenium-python.readthedocs.io/) that can control the browser and make life slightly easier.

**Comment by Zohair-coder** [comment]
After some research and reverse engineering using the browser dev tools, I found that the `shib_idp_session` cookie is responsible for proving the user is authenticated. This cookie is sent only sent on the domain `shib.irttest.drexel.edu`. It is obtained after logging in via Microsoft.

When a request to `https://termmasterschedule.irttest.drexel.edu/webtms_du/` is made, the server checks if a `JSESSIONID` cookie is set or not.

If it is not set:

- a 302 response is sent and we are redirected to `https://shib.irttest.drexel.edu/cas/login?service=https%3A%2F%2Ftermmasterschedule.irttest.drexel.edu%2Fwebtms_du%2Fj_spring_cas_security_check`. In this 302 response, a `Set-Cookie` header is also sent that sets a `JSESSIONID` on the `termmasterschedule.irttest.drexel.edu` domain for future purposes.
- Next, the redirect to `https://shib.irttest.drexel.edu/cas/login?service=https%3A%2F%2Ftermmasterschedule.irttest.drexel.edu%2Fwebtms_du%2Fj_spring_cas_security_check` also returns an HTTP 302 redirect. However, it will redirect to different locations depending on whether or not the `shib_idp_session` cookie is defined or not.
    - If the `shib_idp_session` cookie is not set, it will redirect to `https://shib.irttest.drexel.edu/idp/profile/cas/login?execution=e1s1`, which will return an HTML page that will prompt the user to start the authentication process. Once the user has authenticated, it will send the `Set-Cookie: shib_idp_session=<value>` header to ensure the cookie is present when the user visits the `https://shib.irttest.drexel.edu` domain in the future.
    - If the `shib_idp_session` cookie is set, it will redirect to `https://termmasterschedule.irttest.drexel.edu/webtms_du/j_spring_cas_security_check?ticket=<ticket-value>`, which will then redirect to `
https://termmasterschedule.irttest.drexel.edu/webtms_du/` and show us the actual Drexel Term Master Schedule.

If it is already set:

- The server will use the JSESSIONID to validate whether the user is logged in, and if so, display the actual Drexel Term Master Schedule.

So in summary, there are two cookies that affect the auth flow: the `JSESSIONID`, and the `shib_idp_session`. It is my understanding that the `JSESSIONID` cookie does not last very long - I think it expires within 30 minutes of inactivity but I'm not sure about this. I might test this. On the other hand, I have a **feeling** that the `shib_idp_session` cookie should not expire quickly. This will need to be tested.

I was thinking we could log into the site manually from the browser using our Drexel credentials, and then use the developer tools to get the `shib_idp_session` cookie. We can then set this as an environment variable inside the scraper and modify our code so that it uses this cookie to authenticate. This will be very simply to implement, but it all depends on how long the cookie lasts. If it lasts a long time (say ~3 months), then this is a viable approach. We would just need to update the environment variable after a certain time period to get it to work. Otherwise, another more complex approach will need to be looked at.

I've written a Python script to figure out how long the cookie lasts:

```
import requests
import time

shib_idp_session_cookie = input("Enter the shib_idp_session_cookie from your browser: ") 

session = requests.Session()
session.cookies.set("shib_idp_session", shib_idp_session_cookie)

start_time = time.time()

while True:
    response = session.get("https://termmasterschedule.irttest.drexel.edu/webtms_du/")

    if "Fall Quarter" in response.text:
        print("Cookie is still valid. Checking again in 1 minute...")
    else:
        print("\nCookie has expired or is invalid.")
        end_time = time.time()
        duration_sec = end_time - start_time
        print("The cookie was valid for approximately {:.1f} minutes".format(duration_sec / 60))
        break
    
    time.sleep(60)
```

Since the `requests` library handles redirects automatically, we don't have to worry about much except setting the `shib_idp_session` cookie.

I will report back once this script helps me find out how long a cookie lasts.

**Comment by Zohair-coder** [comment]
One caveat with the script that I am using to test this is that the cookie might not expire because the server might be seeing that requests are continuously coming in from this cookie. It might only expire cookies if it finds that the user has been inactive for a certain time period. If this is the case, it shouldn't be a huge problem since the scraper runs every 20 minutes anyways, so if anything, this will mean that the cookie will last indefinitely (unless there's some other mechanism that we don't know about yet at play as well).

**Comment by Zohair-coder** [comment]
I encountered a very weird issue when I was testing how long it takes for the token to expire.

I wrote the Python script I shared [above](https://github.com/Zohair-coder/drexel-scraper/issues/89#issuecomment-2093543878) on my Mac while connected to the Drexel VPN, and it seemed to be working fine. However, I tried the same thing on Tux (which is connected to the Drexel network) and it didn't work - it redirected me to the "connect" html page where users are prompted to sign in using Microsoft auth, and so it printed:

```
Cookie has expired or is invalid.
The cookie was valid for approximately 0.0 minutes
```

I thought this was very weird so I tried the script on my Mac again and it was again working fine. I then loaded the script on my Raspberry Pi and connected it to the Drexel VPN and tried again - and I got the same error as I did on tux. I checked the Python and requests version and they seem to match - but it doesn't seem to work the same on different machines.

Not sure what's going on yet - needs more investigation.

**Comment by Zohair-coder** [comment]
UPDATE: Drexel IT has moved the SSO requirement to prod, so this issue is more urgent now.

I generated the `shib_idp_session` cookie from my browser and plugged it into the Python script above. It worked for me on my computer. However, if I give someone the same cookie (who are on their own personal computer), and they run the script, the script fails. Drexel's servers are somehow detecting which computer generated the cookie and only giving access to those computers (probably using the IP address but I'm not sure - I tried setting the User-Agent to the same one that my computer sends but it didn't work).
